modules/dnn/src/llm/gguf_parser.cpp

```diff
@@ -0,0 +1,235 @@
+    if (t.size() == 0) {
+        throw std::runtime_error("No input tensors found");
+    }
+    return t[0].dims[0] / 3;
+}
+
+MatShape BlockMetadata::getOutputShape() 
+{
+    std::vector<TensorMetadata> t = getTensorMetadata( tensor_role::attn_qkv, false, false);
+    if (t.size() == 0) {
+        throw std::runtime_error("No input tensors found");
+    }
+
+    MatShape outputShape(3);
+    outputShape[0] = 1; outputShape[1] = -1;
+    outputShape[2] = t[0].dims[1];
+    return outputShape;
+}
+
+void BlockMetadata::getAttentionLayerParams(Ptr<GGUFBufferReader> tensor_reader, LayerParams& layerParams) 
+{
+    layerParams.type = "Attention";
+    layerParams.name = "attention_block";
+    layerParams.set("num_heads", n_heads);
+    layerParams.set("blockn", blockn);
+    int d_hidden = getDhidden();
+    std::vector<int> qkv_hidden_sizes = {d_hidden, d_hidden, d_hidden};
+    layerParams.set("qkv_hidden_sizes", DictValue::arrayInt(&qkv_hidden_sizes[0], 3));
+    Mat qkv_weight = getTensorMetadata(tensor_role::attn_qkv, false, false)[0].getMat(tensor_reader);
+    Mat qkv_bias = getTensorMetadata(tensor_role::attn_qkv, true, false)[0].getMat(tensor_reader);
+    layerParams.blobs.push_back(qkv_weight.t());
+    layerParams.blobs.push_back(qkv_bias);
+}
+
+std::vector<TensorMetadata> BlockMetadata::getTensorMetadata(tensor_role role, bool is_bias, bool allow_multiple) {
+    std::vector<TensorMetadata> result;
+    for (const auto & tensor : tensors) {
+        if (tensor.role == role && tensor.is_bias == is_bias) {
+            result.push_back(tensor);
+        }
+    }
+    return result;
+    if (!allow_multiple && result.size() > 1) {
+        throw std::runtime_error("Multiple tensors found for role: " + std::to_string(role));
+    }
+    return result;
+}
+
+void GGUFParser::addTensorMetadata() {
+    auto tensor = TensorMetadata();
+
+    std::string tensor_name = reader.readString();
+    std::regex layerRegex(R"(^blk\.(\d+)\.([a-zA-Z0-9_]+)\.(weight|bias)$)");
+    std::smatch match;
+    std::string layerName, paramType;
+
+    int blockn;
+
+    if (std::regex_match(tensor_name, match, layerRegex)) {
+        blockn = std::stoi(match[1].str());
+        layerName = match[2].str();
+        paramType = match[3].str();
+        if (paramType == "weight") {
+            tensor.is_bias = false;
+        } else if (paramType == "bias") {
+            tensor.is_bias = true;
+        } else {
+            throw std::runtime_error("Unknown parameter type: " + paramType);
+        }
+        tensor.role = get_tensor_role(layerName);
+        tensor.n_block = blockn;
+    } else {
+        throw std::runtime_error("Invalid tensor name format: " + tensor_name);
+    }
+
+    int dim_count = reader.readSingleValueInt<uint32_t>();
+    tensor.dims = MatShape(dim_count);
+    std::vector<int> dims;
+    for (uint32_t i = 0; i < dim_count; ++i) {
+        dims.push_back(reader.readSingleValueInt<int64_t>());    
+    }
```

Comment:
It's more efficient to use  std::vector<int> dims(dim_count); and assign values, rather than call push_back and trigger reallocations.